File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        counter_t = Counter(t)
        unique_ch = len(counter_t)
        i, j = 0, 0
        mini = float('inf')
        res = []
        while j < len(s):
            if s[j] in counter_t:
                counter_t[s[j]] -= 1
                if counter_t[s[j]] == 0:
                    unique_ch -= 1
            
            while unique_ch == 0 and i <= j:
                if j-i+1 < mini:
                    mini = j-i+1
                    res = [i, j]
                
                if s[i] in counter_t:
                    counter_t[s[i]] += 1
                    if counter_t[s[i]] == 1:
                        unique_ch += 1
                i += 1
            j += 1
        return "" if mini == float('inf') else s[res[0]:res[1]+1]
                
        
        
        

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        int i = 0;
        int j = 0;
        int freq;
        int[] res = new int[] {-1,-1};
        int uniqueCharactersCount = t.length();
        int minWindowSize = Integer.MAX_VALUE;
        char ch;
        char charAtI;
        Map<Character, Integer> tFreqMap = new HashMap<>();
        for (char chT : t.toCharArray()) {
            tFreqMap.put(chT, tFreqMap.getOrDefault(chT, 0) + 1);
        }
        while(j < s.length()) {
            ch = s.charAt(j);
            if(tFreqMap.containsKey(ch)) {
                 tFreqMap.put(ch, tFreqMap.get(ch) - 1);
                if(tFreqMap.get(ch) >= 0) {uniqueCharactersCount--;}
            }
            while(uniqueCharactersCount == 0) {
                if ((j-i+1) < minWindowSize) {
                    minWindowSize = j-i+1;
                    res[0] = i;
                    res[1] = j;
                }
                charAtI = s.charAt(i);
                if (tFreqMap.containsKey(charAtI)) {
                    tFreqMap.put(charAtI, tFreqMap.get(charAtI) + 1);
                    if (tFreqMap.get(charAtI) == 1) ++uniqueCharactersCount;
                }
                ++i;
            }
            ++j;
        }
        
        if (res[0] == -1) {
            return "";
        }
        
        return s.substring(res[0], res[1] + 1);
    }
}